internal/rewrite_action.go

```diff
@@ -58,54 +66,50 @@ func (a *RewriteAction) ToMap() map[string]interface{} {
 	return node
 }
 
-func (a *RewriteAction) Apply(input *chart.Chart) (*chart.Chart, error) {
-	dependencies := input.Dependencies()
-	foundInDependency := false
-	for dependencyIndex, dependency := range dependencies {
-		if dependency.Name() == a.TopLevelKey() {
-			foundInDependency = true
-			valuesIndex, data := GetChartValues(dependency)
-			value := map[string]string{
-				a.GetKey(): a.Value,
-			}
-			newData, err := yamlops2.UpdateMap(data, a.GetSubPathToMap(), "", nil, value)
-			if err != nil {
-				return nil, fmt.Errorf("failed to apply modification to %s: %w", dependency.Name(), err)
-			}
+// Apply will try to execute the rewrite action declaration on the given Helm Chart or sub-charts
+func (a *RewriteAction) Apply(chart *chart.Chart) error {
+	chartToApply, relativeRewriteRule := a.FindChartDestination(chart)
+	return applyUpdate(chartToApply, relativeRewriteRule)
+}
 
-			dependencies[dependencyIndex].Raw[valuesIndex].Data = newData
-		}
+// Apply the yaml update described in the rewrite action to the provided Helm Chart
+func applyUpdate(chart *chart.Chart, a *RewriteAction) error {
+	valuesIndex, data := getChartValues(chart)
+	value := map[string]string{
+		a.GetKey(): a.Value,
 	}
 
-	if foundInDependency {
-		input.SetDependencies(dependencies...)
-	} else {
-		valuesIndex, data := GetChartValues(input)
-		value := map[string]string{
-			a.GetKey(): a.Value,
-		}
-		newData, err := yamlops2.UpdateMap(data, a.GetPathToMap(), "", nil, value)
-		if err != nil {
-			return nil, fmt.Errorf("failed to apply modification to %s: %w", input.Name(), err)
-		}
-
-		input.Raw[valuesIndex].Data = newData
+	newData, err := yamlops2.UpdateMap(data, a.GetPathToMap(), "", nil, value)
+	if err != nil {
+		return fmt.Errorf("failed to apply modification to %s: %w", chart.Name(), err)
 	}
 
-	return input, nil
+	chart.Raw[valuesIndex].Data = newData
+
+	return nil
 }
 
-func (a *RewriteAction) FindChartDestination(parentChart *chart.Chart) *chart.Chart {
+// FindChartDestination will recursively find the Helm Chart a rewrite action will apply to
+// by starting on a parentChart
+// Additionally it will return the rewrite action with path relative to that Helm Chart
+func (a *RewriteAction) FindChartDestination(parentChart *chart.Chart) (*chart.Chart, *RewriteAction) {
```

Comment:
Cannot a rewrite action apply to several charts?
This looks as it only the first match would get selected

Reply:
I don't think it can. The path is the domain including prefixes which might or might not match specific chart names.

having the same chartName at the same level, i.e two mariadb subcharts at the same level is not possible.

Reply:
I see. ok!